test/test_imports.py

```diff
@@ -0,0 +1,64 @@
+import ast
+import os
+from _ast import Import, ImportFrom
+from pathlib import Path
+
+import isort
+import toml
+from pyproject_parser import PyProject
+
+
+def test_for_missing_dependencies() -> None:
+    # All tools out there are too powerful because they find all the imports in the haystack package
+    # We need to find only the top level imports
+
+    #### Collect imports
+    top_level_imports = set()
+    for root, dirs, files in os.walk("haystack"):
+        for file in files:
+            path = Path(root) / file
+
+            if not path.suffix.endswith(".py"):
+                continue
+
+            content = path.read_text(encoding="utf-8")
+            tree = ast.parse(content)
+            for item in tree.body:
+                if isinstance(item, Import):
+                    module = item.names[0].name
+                elif isinstance(item, ImportFrom) and item.level == 0:  # level > 1 are relative imports
+                    module = item.module
+                else:
+                    # we only care about imports
+                    break
```

Comment:
There's a problem with this, we're skipping completely lazy imports. 

We should check for `isinstance(item, With)` too. Should we maybe get only those inside the `LazyImport` context? 🤔

Reply:
Nevermind, just realised this is the scope of the PR. 🤦